src/instructlab/model/phased_training.py

```diff
@@ -0,0 +1,140 @@
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: list[pydantic.DirectoryPath]
+    finished_checkpoints: list[pydantic.DirectoryPath] = []
+    results: list[EvalResult] = []
+    best_checkpoint: EvalResult | None = None
+
+    @pydantic.field_serializer(
+        "checkpoints",
+    )
+    def pathlibPath_list_to_str(self, paths: list[pathlib.Path]) -> list[str]:
+        return [str(path) for path in paths]
+
+
+class TrainPhaseModel(pydantic.BaseModel):
+    """Stores info about training phase. Doesn't store training args because user can get those elsewhere, e.g. config."""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: pydantic.DirectoryPath
+    # TODO: might want training args here, but can't currently do it
+    # without side-effects because model instance is changed inside of _training_phase
+
+
+class JournalModel(pydantic.BaseModel):
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    current_phase: TrainingPhases = TrainingPhases.TRAIN1
+    train_1: TrainPhaseModel | None = None
+    eval_1: EvalPhaseModel | None = None
+    train_2: TrainPhaseModel | None = None
+    eval_2: EvalPhaseModel | None = None
+    final_output: EvalResult | None = None
+
+
+class TrainingJournal:
+    def __init__(self, journalfile: pathlib.Path):
+        self.journalfile = journalfile
+        if journalfile.is_file():
+            if journalfile.stat().st_size == 0:
+                logger.debug("Journalfile is empty- creating a new journalfile.")
+                # if journal is empty, use path but initialize model.
+                self.journal = JournalModel()
+                self.commit()
+            else:
+                logger.debug("Journalfile exists- trying to load.")
+                with open(journalfile, "r") as f:
+                    model_dict = yaml.safe_load(stream=f)
+                    try:
+                        self.journal = JournalModel.parse_obj(model_dict)
+                        logger.debug("Model loaded from Journalfile")
+                    except pydantic.ValidationError:
+                        print(
+                            "Tried to load Journal from Journalfile but encountered an error. Journalfiles must be valid or empty."
+                        )
+                        raise
+        elif journalfile.is_dir():
+            # add 'journalfile.yaml' to the directory by convention
+            logger.debug(
+                "'journalfile' path points to a directory. Creating a new 'journalfile.yaml'"
+            )
+            self.journalfile = journalfile / "journalfile.yaml"
+            self.journal = JournalModel()
+            self.commit()
+        else:
+            raise FileExistsError(f"'journalfile' doesn't exist: {str(journalfile)}")
+
+    def commit(self) -> None:
+        with open(self.journalfile, "w") as f:
```

Comment:
Because it's a journal and consistency is key, the operation should be as atomic as possible. I'd suggest  a few things:

* Use a file-locking mechanism to acquire an exclusive lock before writing to the journal. 
* Release the lock only after the write completes. 
* Flushing buffer and persist data to disk 'immediately'  after write

This should look like this:

```python
import os
import fcntl

def commit(self) -> None:
    with open(self.journalfile, "w") as f:
        # Acquire an exclusive lock to prevent other processes from writing
        fcntl.flock(f, fcntl.LOCK_EX)
        
        # Write journal's content
        yaml.safe_dump(data=self.journal.model_dump(), stream=f)
        
        # Flush the buffer to ensure data is moved to OS buffer
        f.flush()
        
        # Call fsync to ensure the data is physically written to disk
        os.fsync(f.fileno())
        
        # Release the lock (automatically released when 'with' block exits)
```

Reply:
gotcha I'll add that.

Reply:
added